googlemaps/matching-algo.js

```diff
@@ -0,0 +1,152 @@
+      destination: [userTrip.endLocation.coordinates[1],userTrip.endLocation.coordinates[0]],
+      mode: "driving"
+    }
+  ).asPromise()
+  .then((response) => {
+    let userDur = response.json.routes[0].legs[0].duration.value
+    return userDur;
+  })
+  .catch((err) => {
+    console.log(err);
+  });
+};
+
+
+function calcBestMatch(isDriver, userTrip, db){
+
+  let minDur = 99999999;
+  const bestMatch = {};
+
+  //for each trip in the array, we ask google map for the duration of the ABCD trip
+  var googleRequests =
+    db.map(currentTrip => {
+
+      const userStartLat = userTrip.startLocation.coordinates[1];
+      const userStartLong = userTrip.startLocation.coordinates[0];
+
+      const userEndLat = userTrip.endLocation.coordinates[1];
+      const userEndLong = userTrip.endLocation.coordinates[0];
+
+      const currentTripStartLat= currentTrip.startLocation.coordinates[1];
+      const currentTripStartLong = currentTrip.startLocation.coordinates[0];
+
+      const currentTripEndLat= currentTrip.endLocation.coordinates[1];
+      const currentTripEndLong = currentTrip.endLocation.coordinates[0];
+
+      const itinerary = isDriver
+        ? {
+          origin: [userStartLat, userStartLong],
+          destination: [userEndLat, userEndLong],
+          waypoints: [
+            [currentTripStartLat, currentTripStartLong],
+            [currentTripEndLat, currentTripEndLong]
+          ],
+          mode: "driving"
+        } : {
+          origin: [currentTripStartLat, currentTripStartLong],
+          destination: [currentTripEndLat, currentTripEndLong],
+          waypoints: [ 
+            [userStartLat, userStartLong],
+            [userEndLat, userEndLong] 
+          ],
+          mode: "driving"
+        };
+
+      return googleMapsClient.directions(itinerary).asPromise()
+        .then(response => {
+
+          let {legs} = response.json.routes[0]
+
+          //calculating the duration of one ABCD trip (by adding each leg) in seconds
+          //(the response object doesnt give the total duration of a trip but only the duration for each leg)
+          const dur = legs.reduce((duration, currentLeg)=> {
+            return duration += currentLeg.duration.value;
+          }, 0)
+
+          //one we ve got the duration of one ABCD trip, 
+          //we store that trip only if it is shorter than the other ones we've calculated so far
+          if (dur < minDur) {
+            minDur = dur;
+            bestMatch.trip = currentTrip;
+            bestMatch.dur = minDur;
+          }
+
+        })
+        .catch(err => console.log("Error calculating one trip duration", err));
+    });
+
+  return Promise.all(googleRequests)
+    .then(() => bestMatch)
+    .catch(err => console.log("error calculating best match", err));
+}
+
+
+
+//FAKE VARIABLES TO TEST THE FUNCTIONS
+//------------------------------------------------------------------------------------------------------------------
```

Comment:
I think we can move this to another file (ex: matching-algo.test.js) to just keep this nice a clean. Eventually we can add jest to the app for testing - super easy to integrate! :)

Reply:
Yeah, that's definitely something I want to do !! 
I dont have much time this week but hopefully next weekend :)